pkg/reconcilermanager/controllers/rootsync_controller.go

```diff
@@ -779,52 +794,112 @@ func (r *RootSyncReconciler) validateRootSecret(ctx context.Context, rootSync *v
-			"ensuring clusterrolebinding does not exist",
-		)
+	rbList := rbacv1.RoleBindingList{}
+	if err := r.client.List(ctx, &rbList, opts...); err != nil {
+		return nil, errors.Wrap(err, "listing RoleBindings")
 	}
-
-	if roleConfig.Name == "" {
-		roleConfig.Name = "cluster-admin"
+	for idx, rb := range rbList.Items {
+		roleRef := v1beta1.RootSyncRoleRef{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: rb.RoleRef.Kind,
+				Name: rb.RoleRef.Name,
+			},
+			Namespace: rb.Namespace,
+		}
+		currentRoleMap[roleRef] = &rbList.Items[idx]
 	}
-
-	_, err := r.upsertClusterRoleBinding(ctx, reconcilerRef, roleConfig.Name)
-	return errors.Wrap(err, "upserting cluster role binding")
+	crbList := rbacv1.ClusterRoleBindingList{}
+	if err := r.client.List(ctx, &crbList, opts...); err != nil {
+		return nil, errors.Wrap(err, "listing ClusterRoleBindings")
+	}
+	for idx, crb := range crbList.Items {
+		roleRef := v1beta1.RootSyncRoleRef{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: crb.RoleRef.Kind,
+				Name: crb.RoleRef.Name,
+			},
+		}
+		currentRoleMap[roleRef] = &crbList.Items[idx]
+	}
+	return currentRoleMap, nil
 }
 
-func (r *RootSyncReconciler) upsertClusterRoleBinding(ctx context.Context, reconcilerRef types.NamespacedName, clusterRole string) (client.ObjectKey, error) {
-	crbRef := client.ObjectKey{Name: RootSyncPermissionsName(reconcilerRef.Name)}
-	childCRB := &rbacv1.ClusterRoleBinding{}
-	childCRB.Name = crbRef.Name
+func (r *RootSyncReconciler) cleanRBACBindings(ctx context.Context, reconcilerRef types.NamespacedName) error {
+	currentRefMap, err := r.listCurrentRoleRefs(ctx, reconcilerRef)
+	if err != nil {
+		return err
+	}
+	for _, binding := range currentRefMap {
+		if err := r.cleanup(ctx, binding); err != nil {
+			return errors.Wrap(err, "deleting RBAC Binding")
+		}
+	}
+	return nil
+}
 
-	op, err := CreateOrUpdate(ctx, r.client, childCRB, func() error {
-		childCRB.OwnerReferences = nil
-		childCRB.RoleRef = rolereference(clusterRole, "ClusterRole")
-		childCRB.Subjects = []rbacv1.Subject{r.serviceAccountSubject(reconcilerRef)}
-		// Remove existing OwnerReferences, now that we're using finalizers.
-		childCRB.OwnerReferences = nil
-		return nil
-	})
+func (r *RootSyncReconciler) configureClusterRoleBinding(ctx context.Context, reconcilerRef types.NamespacedName, roleRefs []v1beta1.RootSyncRoleRef) error {
+	currentRefMap, err := r.listCurrentRoleRefs(ctx, reconcilerRef)
 	if err != nil {
-		return crbRef, err
+		return err
 	}
-	if op != controllerutil.OperationResultNone {
-		r.logger(ctx).Info("Managed object upsert successful",
-			logFieldObjectRef, crbRef.String(),
-			logFieldObjectKind, "ClusterRoleBinding",
-			logFieldOperation, op)
+	declaredRoleRefs := roleRefs
+	// If no RoleRefs are declared, default to cluster-admin for backwards compatibility
+	if len(declaredRoleRefs) == 0 {
+		declaredRoleRefs = append(declaredRoleRefs, v1beta1.RootSyncRoleRef{
+			RoleRefBase: v1beta1.RoleRefBase{
```

Comment:
One thing to note is that while this backwards-compatible default does keep the `cluster-admin` default ClusterRoleBinding, the name/structure of the ClusterRoleBinding does change. Instead of a many RootSync -> one ClusterRoleBinding named `configsync.gke.io:root-reconciler`, it's now a many->many ClusterRoleBinding where each RootSync will get it's own ClusterRoleBinding with a generated name.

This assumes that the name/structure of the ClusterRoleBinding is an implementation detail and therefore changes are not breaking. We could alternatively maintain a separate code path for the legacy behavior of a singleton `configsync.gke.io:root-reconciler` ClusterRoleBinding with many subjects, which IMO has the downside of more code variance/complexity.

Reply:
Our general policy is simply not to require customers to need to do any work when a service is auto-upgraded, such as Config Sync on Config Controller. But we don't have any way to determine if any customers depend on the exact result of the generated resources. So historically we have not considered the generated resources to be a customer facing interface.

We don't want to leak any resources, tho. So if we change the naming scheme, we need to clean up the legacy resources. This can be done in the config-management-operator on upgrade with a version check or in the reconciler-manager teardown code.

Reply:
You might also consider de-duping similar ClusterRoleBindings and RoleBindings, to avoid sprawl.

Reply:
The legacy resources created by older versions reconciler-manager are already being cleaned up by the current reconciler-manager, see below.

---

De-duping may work for the ClusterRoleBindings/RoleBindings with known names, such as the binding to the "base" ClusterRole. However it's less clear to me how we can handle de-duping for user-provided roleRefs - given we need to maintain an inventory of previously applied bindings.

In the current approach, each binding is attached to a specific reconciler and can be listed using a label selector. If we consolidate bindings and attach multiple reconcilers as subjects, it becomes less clear to me how we can track which bindings were attached to a given reconciler on previous iterations.

Solutions that come to mind but don't scale:
- list all config sync bindings and filter the subjects in code
- attach a label for each reconciler subject to the binding

Reply:
I think the only de-duping we might want would be a single admin ClusterRoleBinding, and just add a new service account to it for every RootSync using the default config. Then if a custom config is specified, use reconciler-specific ClusterRoleBindings/RoleBindings.

Reply:
For reference, the migration cleanup implementation in #938 is [here](https://github.com/GoogleContainerTools/kpt-config-sync/blob/3a526f88a0a978a84c6b559ba506d47e582e1c35/pkg/reconcilermanager/controllers/rootsync_controller.go#L783-L788).

Reply:
I kept that snippet but moved it to the end of the function, the idea being that we should add new bindings before removing old ones

Reply:
@karlkfi updated to use a consolidated binding for both the base bindings and for the cluster-admin binding